### PR TITLE
Fixed #33348 -- Changed SimpleTestCase.assertFormError()/assertFormsetErrors() to take form/formset.

### PR DESCRIPTION
**Pull request by bmispelon**
ticket-33348

### CODE REVIEW

File: docs/releases/4.1.txt [hunk 1]
```diff
@@ -323,6 +323,17 @@ Miscellaneous
 
 * ``CSRF_COOKIE_MASKED`` transitional setting is deprecated.
 
+* The API of :meth:`~django.test.SimpleTestCase.assertFormError()` and
+  :meth:`~django.test.SimpleTestCase.assertFormsetError()` have changed.
```
Comment (timgraham) [inline docs/releases/4.1.txt:327]: Write in present tense ("is" instead of "haven been", "was", etc.)

Comment (bmispelon) [reply docs/releases/4.1.txt]: Thanks 👍🏻

**Comment by felixxm** [comment]
I merged the first two commits (with small edits) in 61b332499d28afe1c7d413411166ad405663e750 and e95e6425ac354b8fb0c24580b740dbdc2ea72bd6.

File: django/test/testcases.py [hunk 2]
```diff
@@ -54,9 +57,7 @@ def to_list(value):
     Put value into a list if it's not already one. Return an empty list if
     value is None.
     """
-    if value is None:
-        value = []
-    elif not isinstance(value, list):
```
Comment (felixxm) [inline django/test/testcases.py:59]: I'm not sure about these changes, they seem unnecessary. We can raise a deprecation warning when `None` is passed in `errors` even without them.

File: django/test/testcases.py [hunk 3]
```diff
@@ -488,142 +589,94 @@ def _check_test_client_response(self, response, attribute, method_name):
...
         """
-        self._check_test_client_response(response, 'context', 'assertFormsetError')
-        # Add punctuation to msg_prefix
+        if errors is None:
+            warnings.warn(
+                "Passing errors=None to assertFormError() is deprecated. Use errors=[] instead.",
+                RemovedInDjango50Warning,
+                stacklevel=2
+            )
+            errors = []
```
Comment (felixxm) [inline django/test/testcases.py:627]: Can we deprecate passing `None` in `errors` in a separate PR? and do this in advance. There is already a lot of changes in this patch, that should make it more readable and easier to review. I can try to refactor this out if you don't have time to keep working on this.

**Review by felixxm — COMMENTED** [review]
@bmispelon Thanks :+1:

**Comment by felixxm** [comment]
I resolved conflicts and rebased from the `main` branch.

File: docs/internals/deprecation.txt [hunk 4]
```diff
@@ -62,6 +62,14 @@ details on these changes.
...
   an ``ErrorDict`` and ``ErrorList`` will be removed.
 
+* :meth:`~django.test.SimpleTestCase.assertFormError()` (and
+  :meth:`~django.test.SimpleTestCase.assertFormsetError()`) will no longer accept
+  a response object and a form name but only a form object (resp. formset).
+
+* Passing ``errors=None`` to :meth:`~django.test.SimpleTestCase.assertFormError()`
+  or :meth:`~django.test.SimpleTestCase.assertFormsetError()` will no longer be
+  allowed (use ``errors=[]`` instead).
+
```
Comment (felixxm) [inline docs/internals/deprecation.txt:72]: These should be in features deprecated in Django 4.1.

File: django/test/testcases.py (hunk 3 above)
Comment (bmispelon) [reply django/test/testcases.py:627]: That's fair, I can rework the patch and open a separate ticket for deprecating `errors=None`.

Comment (felixxm) [reply django/test/testcases.py:627]: Thanks :+1: , IMO a separate ticket is not necessary, we can `Refs #33348 -- ...`.

Comment (bmispelon) [reply django/test/testcases.py:627]: I've opened a new PR with only the `errors=None` deprecation: #15282

I will put this current PR on hold until this new one is merged.

**Comment by bmispelon** [comment]
I tried my best but the diff is still a bit hard to read, sorry. If you have some ideas on how to make the review easier I'm all ears.

Meanwhile, here are some explanations of the approach I took.

(to save some typing, any mention of `assertFormError` also applies to `assertFormsetError`)

To simplify the removal of deprecated code, all the legacy logic is moved out of `SimpleTestCase`.
`SimpleTestCase.assertFormError()` uses the new signature (actual object vs response object) and the new logic (exact error comparison instead of subset).

The old `assertFormError(response, form_name, ...)` logic was moved to a separate class (`_assert_form_error_deprecation_helper`) and rewritten to use the new `assertFormError(form_instance, ...)` logic. Essentially all that remains is the "find form instance in context" logic.

Backwards compatibility is maintained via a decorator (`@patch_signature`) on `SimpleTestCase.assertFormError()`.
This decorator uses the `inspect` module to determine if the user is calling the method with the old signature or the new one.
Doing it this way means that any valid combination of positional and keyword arguments is supported and will correctly be dispatched to either the old method or the new one.


The tests in `client_regress` have been left as-is (with the error messages tweaked to match the new format) and provide a strong guarantee that backwards-compatibility is maintained.
The tests in `test_utils.AssertFormErrorDeprecationTests` exhaustively try all possible combinations of positional and keyword arguments, for both the old and the new signature. Again, this provides a strong guarantee that the complex `@patch_signature` decorator works as intended.


I've marked everything that should be deleted post-deprecation with comments (even some views in `test_client_regress` that are only used for legacy `assertFormError` tests).

**Comment by felixxm** [comment]
@smithdc1 Do you have time to keep working on this?

**Comment by felixxm** [comment]
@bmispelon Can you rebase from the `main` branch and apply `black` :gift: ? I should be able to review it in next week.

**Comment by bmispelon** [comment]
> @bmispelon Can you rebase from the `main` branch and apply `black` gift ? I should be able to review it in next week.

Done I think. I'm not sure if I did it wrong, but I had to re-wrap a bunch of strings to fit the 88 character line limit manually (I thought `black` would do it automatically but I couldn't get it to work).

**Comment by felixxm** [comment]
buildbot, test this please.

File: django/contrib/admin/helpers.py [hunk 5]
```diff
@@ -433,9 +441,16 @@ def inline_formset_data(self):
     def forms(self):
         return self.formset.forms
 
-    @property
     def non_form_errors(self):
-        return self.formset.non_form_errors
+        return self.formset.non_form_errors()
```
Comment (felixxm) [inline django/contrib/admin/helpers.py:445]: `InlineAdminFormSet.non_form_errors` is undocumented. Nevertheless, I'd add a short release note in breaking change, e.g.
```
* The undocumented ``InlineAdminFormSet.non_form_errors`` property is replaced
  by the ``non_form_errors()`` method. This is consistent with ``BaseFormSet``.
```

File: django/test/testcases.py [hunk 6]
```diff
@@ -165,6 +166,129 @@ def __call__(self):
         raise DatabaseOperationForbidden(self.message)
 
 
+# RemovedInDjango50Warning (the whole class)
```
Comment (felixxm) [inline django/test/testcases.py:169]:
```suggestion
# RemovedInDjango50Warning
```

File: django/test/testcases.py [hunk 7]
```diff
@@ -165,6 +166,129 @@ def __call__(self):
...
+class _assert_form_error_deprecation_helper:
```
Comment (felixxm) [inline django/test/testcases.py:170]:
```suggestion
class _AssertFormErrorDeprecationHelper:
```

File: django/test/testcases.py [hunk 8]
```diff
@@ -563,186 +687,109 @@ def _check_test_client_response(self, response, attribute, method_name):
...
-                        "contain the non-form error '%s' (actual errors: %s)"
-                        % (formset, i, err, repr(non_form_errors)),
-                    )
-        if not found_formset:
+        if not formset.is_bound:
             self.fail(
                 msg_prefix
-                + "The formset '%s' was not used to render the response" % formset
+                + f"The formset {formset!r} is not bound, it will never have any errors"
+            )
```
Comment (felixxm) [inline django/test/testcases.py:775]: I moved this an a similar check for forms to the separate PR, see #15425.

**Review by felixxm — COMMENTED** [review]
@bmispelon Many thanks :+1: I will review remaining code after merging #15425.

File: django/test/testcases.py [hunk 9]
```diff
@@ -563,198 +687,108 @@ def _check_test_client_response(self, response, attribute, method_name):
...
-                + "The formset '%s' was not used to render the response" % formset
+                f"{msg_prefix}The formset {formset!r} is not bound, it will never have "
+                f"any errors."
+            )
+        if form_index is not None and form_index >= formset.total_form_count():
+            form_count = formset.total_form_count()
+            failure_message = "The formset {!r} only has {} form{}".format(
+                formset, form_count, "s" if form_count > 1 else ""
+            )
+            self.fail(msg_prefix + failure_message)
```
Comment (felixxm) [inline django/test/testcases.py:781]: I moved this to a separate PR, see #15426.

File: tests/admin_views/tests.py [hunk 10]
```diff
@@ -6689,10 +6691,9 @@ def test_password_mismatch(self):
             },
         )
         self.assertEqual(response.status_code, 200)
-        self.assertFormError(response, "adminform", "password", [])
+        self.assertFormError(response.context["adminform"], "password1", [])
```
Comment (felixxm) [inline tests/admin_views/tests.py:6694]: I moved this to a separate PR, see #15426.

File: tests/admin_views/tests.py [hunk 11]
```diff
@@ -7814,32 +7815,23 @@ def test_add_view_form_and_formsets_run_validation(self):
...
-        msg = (
-            "The form 'adminform' in context 0 does not contain the non-field error "
-            "'Error'"
+            response.context["adminform"],
+            "some_required_info",
+            ["This field is required."],
         )
-        with self.assertRaisesMessage(AssertionError, msg):
-            self.assertFormError(response, "adminform", None, ["Error"])
+        self.assertFormError(response.context["adminform"], None, [])
```
Comment (felixxm) [inline tests/admin_views/tests.py:7822]: I moved this to a separate PR, see #15426.

File: tests/admin_views/tests.py [hunk 12]
```diff
@@ -7814,32 +7815,23 @@ def test_add_view_form_and_formsets_run_validation(self):
...
             [
                 "Children must share a family name with their parents in this "
                 "contrived test case"
             ],
         )
-        msg = (
-            "The formset 'inline_admin_formset' in context 22 does not contain any "
-            "non-form errors."
+        self.assertFormsetError(
+            response.context["inline_admin_formset"], None, None, []
```
Comment (felixxm) [inline tests/admin_views/tests.py:7833]: I moved this to a separate PR, see #15426.

**Review by felixxm — COMMENTED** [review]
@bmispelon I moved three more changes to a separate PR, see #15426.

I have one more locally, that makes raised errors context-free and introduces `_assert_form_error()` helper. This is probably the last cleanup before the main change :relieved:

**Comment by felixxm** [comment]
> I have one more locally, that makes raised errors context-free and introduces `_assert_form_error()` helper. This is probably the last cleanup before the main change relieved

@bmispelon I refactored out the last (probably) cleanup, see #15427.

File: django/test/testcases.py [hunk 13]
```diff
@@ -658,50 +755,36 @@ def assertFormsetError(
                 stacklevel=2,
             )
             errors = []
-        # Put error(s) into a list to simplify processing.
+
+        if form_index is None and field is not None:
+            raise TypeError("You must use field=None when using formset=None")
```
Comment (felixxm) [inline django/test/testcases.py:760]:
```suggestion
            raise TypeError("You must use field=None when using form_index=None")
```

Comment (felixxm) [reply django/test/testcases.py:760]: I fixed this in a separate PR, see #15428.

File: tests/test_client_regress/tests.py [hunk 14]
```diff
@@ -739,6 +744,9 @@ def test_unknown_nonfield_error(self):
             )
 
 
+# RemovedInDjango50Warning (whole TestCase): relevant tests are now in
+# test_utils.tests.AssertFormsetErrorTests
+@ignore_warnings(category=RemovedInDjango50Warning)
```
Comment (felixxm) [inline tests/test_client_regress/tests.py:749]: @bmispelon Is there any value in keeping these tests if they are already redundant? What do you think about adding missing tests to `test_utils.tests` (if any) and removing `AssertFormsetErrorTests` and `AssertFormErrorTests` from `test_client_regress` in a separate commit?

**Comment by felixxm** [comment]
@bmispelon Do you have time to keep working on this?

File: docs/topics/testing/tools.txt [hunk 15]
```diff
@@ -1473,47 +1473,64 @@ your test suite.
...
-    ``form`` is the name the ``Form`` instance was given in the template
-    context of the response.
+    .. versionchanged:: 4.1
 
-    ``field`` is the name of the field on the form to check. If ``field``
-    has a value of ``None``, non-field errors (errors you can access via
-    :meth:`form.non_field_errors() <django.forms.Form.non_field_errors>`) will
-    be checked.
+        In older versions, using an empty error list with ``assertFormError()``
+        always pass, regardless of whether the field had any errors or not.
```
Comment (carltongibson) [inline docs/topics/testing/tools.txt:1497]: `will always pass`, or `would always pass`, or `always passed` — I would favour `would always pass`, which has a sense of `not current`.

File: docs/topics/testing/tools.txt [hunk 16]
```diff
@@ -1473,47 +1473,64 @@ your test suite.
...
+        Starting from Django 4.1, using ``errors=[]`` will only pass if the
+        field actually has no errors.
 
-    ``errors`` is an error string, or a list of error strings, that are
-    expected as a result of form validation.
+        Django 4.1 also changed the behavior of ``assertFormError()`` when a
+        field has multiple errors. In older version, if a field had multiple
+        errors and you'd check for only some of them the test would pass.
```
Comment (carltongibson) [inline docs/topics/testing/tools.txt:1503]:
```suggestion
        field has multiple errors. In older versions, if a field had multiple
        errors and you checked for only some of them, the test would pass.
```

File: docs/topics/testing/tools.txt [hunk 17]
```diff
@@ -1473,47 +1473,64 @@ your test suite.
...
+        Starting from Django 4.1, the error list must be an exact match to the
+        field's actual errors.
 
-.. method:: SimpleTestCase.assertFormsetError(response, formset, form_index, field, errors, msg_prefix='')
+    .. deprecated:: 4.1
+
+        Support for passing a response object and a form name to
+        ``assertFormError()`` is deprecated and will be removed in Django 5.0.
```
Comment (carltongibson) [inline docs/topics/testing/tools.txt:1510]: Maybe worth repeating the `Use the form instance directly instead` bit. 🤔

**Review by carltongibson — COMMENTED** [review]
Looking good! 👍 — I just left a couple of comments on the docs changes, which I think read quite well. 

Thanks all!

**Comment by felixxm** [comment]
@carltongibson Thanks :+1: Docs updated.
